File: dashboards/ruby/routes/agents.rb
# frozen_string_literal: true

# Agents routes — list, create, detail, archive, actions

get '/agents' do
  require_auth!
  res = api_get('/api/agents')
  @agents = res[:status] == 200 ? (res[:body]['agents'] || res[:body]) : []
  @agents = @agents.is_a?(Array) ? @agents : []
  erb :agents
end

get '/agents/:id' do
  require_auth!
  res = api_get("/api/agents/#{params[:id]}")
  halt 404, erb(:layout) { 'Agent not found' } unless res[:status] == 200
  @agent = res[:body].is_a?(Hash) ? res[:body] : (res[:body]['agent'] || {})
  @config = @agent['config'] || {}

  # Fetch activity data (events, tool calls, journal)
  agent_id = params[:id]
  begin
    ev_res = api_get("/activity/events?agentId=#{agent_id}&limit=50")
    @events = ev_res[:status] == 200 ? (ev_res[:body]['events'] || ev_res[:body]) : []
    @events = @events.is_a?(Array) ? @events : []
  rescue => e
    @events = []
  end

  begin
    tc_res = api_get("/activity/tool-calls?agentId=#{agent_id}&limit=50")
    @tool_calls = tc_res[:status] == 200 ? (tc_res[:body]['toolCalls'] || tc_res[:body]['tool_calls'] || tc_res[:body]) : []
    @tool_calls = @tool_calls.is_a?(Array) ? @tool_calls : []
  rescue => e
    @tool_calls = []
  end

  begin
    j_res = api_get("/engine/journal?agentId=#{agent_id}&orgId=default&limit=50")
    @journal_entries = j_res[:status] == 200 ? (j_res[:body]['entries'] || j_res[:body]['journal'] || j_res[:body]) : []
    @journal_entries = @journal_entries.is_a?(Array) ? @journal_entries : []
  rescue => e
    @journal_entries = []
  end

  # Fetch agent-level tool security
  begin
    ts_res = api_get("/engine/agents/#{agent_id}/tool-security")
    if ts_res[:status] == 200
      @agent_tool_security = ts_res[:body]['toolSecurity'] || {}
      @agent_tool_security_org = ts_res[:body]['orgDefaults'] || {}
      @agent_tool_security_overrides = ts_res[:body]['agentOverrides'] || {}
    else
      @agent_tool_security = {}
      @agent_tool_security_org = {}
      @agent_tool_security_overrides = {}
    end
  rescue => e
    @agent_tool_security = {}
    @agent_tool_security_org = {}
    @agent_tool_security_overrides = {}
  end

  erb :agent_detail
end

post '/agents/:id/deploy' do
  require_auth!
  res = api_post("/api/agents/#{params[:id]}/deploy", {})
  if res[:status] < 300
    set_flash('Agent deployed successfully', 'success')
  else
    set_flash(res[:body].is_a?(Hash) ? (res[:body]['error'] || 'Failed to deploy agent') : 'Failed to deploy agent', 'danger')
  end
  redirect "/agents/#{params[:id]}"
end

post '/agents/:id/stop' do
  require_auth!
  res = api_post("/api/agents/#{params[:id]}/stop", {})
  if res[:status] < 300
    set_flash('Agent stopped successfully', 'success')
  else
    set_flash(res[:body].is_a?(Hash) ? (res[:body]['error'] || 'Failed to stop agent') : 'Failed to stop agent', 'danger')
  end
  redirect "/agents/#{params[:id]}"
end

post '/agents/:id/restart' do
  require_auth!
  res = api_post("/api/agents/#{params[:id]}/restart", {})
  if res[:status] < 300
    set_flash('Agent restarted successfully', 'success')
  else
    set_flash(res[:body].is_a?(Hash) ? (res[:body]['error'] || 'Failed to restart agent') : 'Failed to restart agent', 'danger')
  end
  redirect "/agents/#{params[:id]}"
end

post '/agents' do
  require_auth!
  payload = {
    name: params[:name],
    description: params[:description],
    provider: params[:provider] || 'anthropic',
    model: params[:model],
    persona: {
      gender: params[:gender].to_s.empty? ? nil : params[:gender],
      dateOfBirth: params[:date_of_birth].to_s.empty? ? nil : params[:date_of_birth],
      maritalStatus: params[:marital_status].to_s.empty? ? nil : params[:marital_status],
      culturalBackground: params[:cultural_background].to_s.empty? ? nil : params[:cultural_background],
      language: params[:language].to_s.empty? ? nil : params[:language],
      traits: {
        communication: params[:trait_communication] || 'direct',
        detail: params[:trait_detail] || 'detail-oriented',
        energy: params[:trait_energy] || 'calm',
        humor: params[:humor] || 'warm',
        formality: params[:formality] || 'adaptive',
        empathy: params[:empathy] || 'moderate',
        patience: params[:patience] || 'patient',
        creativity: params[:creativity] || 'creative',
      },
    },
  }
  payload[:soul_id] = params[:soul_id] unless params[:soul_id].to_s.empty?
  res = api_post('/api/agents', payload)
  if res[:status] < 300
    set_flash('Agent created successfully', 'success')
  else
    set_flash(res[:body].is_a?(Hash) ? (res[:body]['error'] || 'Failed to create agent') : 'Failed to create agent', 'danger')
  end
  redirect '/agents'
end

post '/agents/:id/tool-security' do
  require_auth!
  agent_id = params[:id]

  security = {
    'pathSandbox' => {
      'enabled' => params['path_sandbox_enabled'] == 'on',
      'allowedDirs' => (params['path_sandbox_allowed_dirs'] || '').split(',').map(&:strip).reject(&:empty?),
      'blockedPatterns' => (params['path_sandbox_blocked_patterns'] || '').split(',').map(&:strip).reject(&:empty?)
    },
    'ssrf' => {
      'enabled' => params['ssrf_enabled'] == 'on',
      'allowedHosts' => (params['ssrf_allowed_hosts'] || '').split(',').map(&:strip).reject(&:empty?),
      'blockedCidrs' => (params['ssrf_blocked_cidrs'] || '').split(',').map(&:strip).reject(&:empty?)
    },
    'commandSanitizer' => {
      'enabled' => params['command_sanitizer_enabled'] == 'on',
      'mode' => params['command_sanitizer_mode'] || 'blocklist',
      'allowedCommands' => (params['command_sanitizer_allowed'] || '').split(',').map(&:strip).reject(&:empty?),
      'blockedPatterns' => (params['command_sanitizer_blocked'] || '').split(',').map(&:strip).reject(&:empty?)
    }
  }

  middleware = {
    'audit' => {
      'enabled' => params['audit_enabled'] == 'on',
      'redactKeys' => (params['audit_redact_keys'] || '').split(',').map(&:strip).reject(&:empty?)
    },
    'rateLimit' => {
      'enabled' => params['rate_limit_mw_enabled'] == 'on',
      'overrides' => {}
    },
    'circuitBreaker' => {
      'enabled' => params['circuit_breaker_enabled'] == 'on'
    },
    'telemetry' => {
      'enabled' => params['telemetry_enabled'] == 'on'
    }
  }

  payload = {
    'toolSecurity' => { 'security' => security, 'middleware' => middleware },
    'updatedBy' => 'dashboard'
  }
  res = api_patch("/engine/agents/#{agent_id}/tool-security", payload)
  if res[:status] < 300
    set_flash('Agent tool security updated', 'success')
  else
    set_flash(res[:body].is_a?(Hash) ? (res[:body]['error'] || 'Failed to update tool security') : 'Failed to update tool security', 'danger')
  end
  redirect "/agents/#{agent_id}"
end

post '/agents/:id/archive' do
  require_auth!
  res = api_patch("/api/agents/#{params[:id]}", { status: 'archived' })
  if res[:status] < 300
    set_flash('Agent archived successfully', 'success')
  else
    set_flash(res[:body].is_a?(Hash) ? (res[:body]['error'] || 'Failed to archive agent') : 'Failed to archive agent', 'danger')
  end
  redirect '/agents'
end

File: dashboards/python/routes/agents.py
"""
Agents routes: list, create, detail, and archive AI agents.
"""

import re

from flask import Blueprint, render_template, request, session, redirect, url_for, flash
from functools import wraps
from utils.api import api_request

agents_bp = Blueprint('agents_bp', __name__)


def login_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        if 'token' not in session:
            return redirect(url_for('auth_bp.login'))
        return f(*args, **kwargs)
    return decorated


def _resolve_agent_fields(agent):
    """Resolve display name, email, and model from agent config following priority rules."""
    config = agent.get('config') or {}
    identity = config.get('identity') or {}

    # Name: config.identity.name > config.name > config.displayName > agent.name
    display_name = (
        identity.get('name')
        or config.get('name')
        or config.get('displayName')
        or agent.get('name', '')
    )

    # Email: config.identity.email > config.email > agent.email (never show raw UUIDs)
    email = (
        identity.get('email')
        or config.get('email')
        or agent.get('email', '')
    )

    # Filter out UUID emails
    if email and re.match(r'^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$', email, re.IGNORECASE):
        email = ''

    # Avatar initial: first letter of resolved display name
    avatar_initial = display_name[0].upper() if display_name else '?'

    # Model: if dict, use modelId or provider; otherwise use as-is
    model_raw = config.get('model', '')
    if isinstance(model_raw, dict):
        model = model_raw.get('modelId') or model_raw.get('provider') or ''
    else:
        model = model_raw

    return display_name, email, avatar_initial, model


@agents_bp.route('/agents')
@login_required
def agents():
    data = api_request('GET', '/api/agents')
    return render_template(
        'agents.html',
        active_page='agents',
        user=session.get('user', {}),
        agents=data.get('agents', []),
    )


def _parse_comma_list(value):
    """Parse a comma-separated string into a list, stripping whitespace."""
    if not value:
        return []
    return [item.strip() for item in value.split(',') if item.strip()]


@agents_bp.route('/api/providers/<provider_id>/models')
@login_required
def provider_models(provider_id):
    from flask import jsonify
    data = api_request('GET', f'/api/providers/{provider_id}/models')
    return jsonify(data)


@agents_bp.route('/agents/<id>')
@login_required
def agent_detail(id):
    data = api_request('GET', f'/api/agents/{id}')
    agent = data.get('agent') or data

    display_name, email, avatar_initial, model = _resolve_agent_fields(agent)

    config = agent.get('config') or {}
    persona = config.get('persona') or {}
    traits = persona.get('traits') or {}
    permissions = config.get('permissions') or agent.get('permissions') or {}

    # Fetch activity data (events, tool calls, journal) — each wrapped in try/except
    events = []
    try:
        ev_data = api_request('GET', f'/activity/events?agentId={id}&limit=50')
        events = ev_data.get('events', [])
    except Exception:
        pass

    tool_calls = []
    try:
        tc_data = api_request('GET', f'/activity/tool-calls?agentId={id}&limit=50')
        tool_calls = tc_data.get('toolCalls', [])
    except Exception:
        pass

    journal_entries = []
    try:
        j_data = api_request('GET', f'/journal?agentId={id}&orgId=default&limit=50')
        journal_entries = j_data.get('entries', [])
    except Exception:
        pass

    # Fetch tool security data
    tool_security = {}
    org_defaults = {}
    agent_overrides = {}
    try:
        ts_data = api_request('GET', f'/engine/agents/{id}/tool-security')
        tool_security = ts_data.get('toolSecurity') or {}
        org_defaults = ts_data.get('orgDefaults') or {}
        agent_overrides = ts_data.get('agentOverrides') or {}
    except Exception:
        pass

    return render_template(
        'agent_detail.html',
        active_page='agents',
        user=session.get('user', {}),
        agent=agent,
        display_name=display_name,
        email=email,
        avatar_initial=avatar_initial,
        model=model,
        config=config,
        persona=persona,
        traits=traits,
        permissions=permissions,
        profile=permissions,
        events=events,
        tool_calls=tool_calls,
        journal_entries=journal_entries,
        tool_security=tool_security,
        org_defaults=org_defaults,
        agent_overrides=agent_overrides,
    )


@agents_bp.route('/agents/<id>/tool-security', methods=['POST'])
@login_required
def save_agent_tool_security(id):
    tool_sec = {
        'security': {
            'pathSandbox': {
                'enabled': 'ps_enabled' in request.form,
                'allowedDirs': _parse_comma_list(request.form.get('ps_allowedDirs', '')),
                'blockedPatterns': _parse_comma_list(request.form.get('ps_blockedPatterns', '')),
            },
            'ssrf': {
                'enabled': 'ssrf_enabled' in request.form,
                'allowedHosts': _parse_comma_list(request.form.get('ssrf_allowedHosts', '')),
                'blockedCidrs': _parse_comma_list(request.form.get('ssrf_blockedCidrs', '')),
            },
            'commandSanitizer': {
                'enabled': 'cs_enabled' in request.form,
                'mode': request.form.get('cs_mode', 'blocklist'),
                'allowedCommands': _parse_comma_list(request.form.get('cs_allowedCommands', '')),
                'blockedPatterns': _parse_comma_list(request.form.get('cs_blockedPatterns', '')),
            },
        },
        'middleware': {
            'audit': {
                'enabled': 'audit_enabled' in request.form,
                'redactKeys': _parse_comma_list(request.form.get('audit_redactKeys', '')),
            },
            'rateLimit': {
                'enabled': 'rl_enabled' in request.form,
                'overrides': {},
            },
            'circuitBreaker': {
                'enabled': 'cb_enabled' in request.form,
            },
            'telemetry': {
                'enabled': 'tel_enabled' in request.form,
            },
        },
    }
    result = api_request('PATCH', f'/engine/agents/{id}/tool-security', {
        'toolSecurity': tool_sec,
        'updatedBy': 'dashboard',
    })
    if 'error' in result:
        flash(result.get('error', 'Failed to save tool security'))
    else:
        flash('Agent tool security saved!')
    return redirect(url_for('agents_bp.agent_detail', id=id))


@agents_bp.route('/agents/<id>/tool-security/reset', methods=['POST'])
@login_required
def reset_agent_tool_security(id):
    result = api_request('PATCH', f'/engine/agents/{id}/tool-security', {
        'toolSecurity': {},
        'updatedBy': 'dashboard',
    })
    if 'error' in result:
        flash(result.get('error', 'Failed to reset tool security'))
    else:
        flash('Tool security reset to org defaults')
    return redirect(url_for('agents_bp.agent_detail', id=id))


@agents_bp.route('/agents/<id>/deploy', methods=['POST'])
@login_required
def deploy_agent(id):
    api_request('POST', f'/api/agents/{id}/deploy')
    flash('Agent deployed')
    return redirect(url_for('agents_bp.agent_detail', id=id))


@agents_bp.route('/agents/<id>/stop', methods=['POST'])
@login_required
def stop_agent(id):
    api_request('POST', f'/api/agents/{id}/stop')
    flash('Agent stopped')
    return redirect(url_for('agents_bp.agent_detail', id=id))


@agents_bp.route('/agents/<id>/restart', methods=['POST'])
@login_required
def restart_agent(id):
    api_request('POST', f'/api/agents/{id}/restart')
    flash('Agent restarted')
    return redirect(url_for('agents_bp.agent_detail', id=id))


@agents_bp.route('/agents/create', methods=['POST'])
@login_required
def create_agent():
    body = {
        'name': request.form['name'],
        'role': request.form.get('role', 'assistant'),
        'provider': request.form.get('provider', 'anthropic'),
        'model': request.form.get('model') or None,
        'persona': {
            'gender': request.form.get('gender') or None,
            'dateOfBirth': request.form.get('date_of_birth') or None,
            'maritalStatus': request.form.get('marital_status') or None,
            'culturalBackground': request.form.get('cultural_background') or None,
            'language': request.form.get('language') or None,
            'traits': {
                'communication': request.form.get('trait_communication', 'direct'),
                'detail': request.form.get('trait_detail', 'detail-oriented'),
                'energy': request.form.get('trait_energy', 'calm'),
                'humor': request.form.get('humor') or 'warm',
                'formality': request.form.get('formality') or 'adaptive',
                'empathy': request.form.get('empathy') or 'moderate',
                'patience': request.form.get('patience') or 'patient',
                'creativity': request.form.get('creativity') or 'creative',
            },
        },
    }
    if request.form.get('soul_id'):
        body['soul_id'] = request.form['soul_id']
    if request.form.get('email'):
        body['email'] = request.form['email']
    result = api_request('POST', '/api/agents', body)
    flash('Agent created!' if 'id' in result else result.get('error', 'Failed'))
    return redirect(url_for('agents_bp.agents'))


@agents_bp.route('/agents/<id>/journal/<entry_id>/rollback', methods=['POST'])
@login_required
def rollback_journal(id, entry_id):
    result = api_request('POST', f'/journal/{entry_id}/rollback', body={})
    if result.get('success'):
        flash('Journal entry rolled back')
    else:
        flash(result.get('error', 'Rollback failed'))
    return redirect(url_for('agents_bp.agent_detail', id=id))


@agents_bp.route('/agents/<id>/archive', methods=['POST'])
@login_required
def archive_agent(id):
    api_request('POST', f'/api/agents/{id}/archive')
    flash('Agent archived')
    return redirect(url_for('agents_bp.agents'))
